README
Linux port of DTrace

Mar 2015
Paul D. Fox
paul.d.fox@gmail.com
http://www.twitter.com/crispeditor
http://www.crisp.demon.co.uk

Blog - latest news and stuff about the dtrace project:

   http://crtags.blogspot.com/
   http://www.crisp.demon.co.uk/blog/

Download dtrace tarballs for linux here:

   https://github.com/dtrace4linux/linux
   ftp://crisp.publicvm.com/pub/release/website/dtrace

Introduction
============

This is a port of the Sun DTrace user and kernel code to Linux.
No  linux  kernel  code  is  touched in this build, but what is
produced  is  a dynamically loadable kernel module. This avoids
licensing issues and allows people to load and update dtrace as
they desire.

The  goal  of  this project is to make available DTrace for the
Linux  platforms. By making it available for everyone, they can
use  it  to  optimise their systems and tools, and in return, I
get to benefit from their work.

PayPal
======
If  you want to make a donation for this software, feel free to
do so. Nothing is asked of you - it is genuinely free software,
but it can help guage interest and appreciation if you do.

You  can  pay  by  visiting  the  link  below  and  clicking on
"Donate", or use this reference for donations:

   foxpaypal@crisp.demon.co.uk

Licensing
=========

The  original  DTrace is licensed under Sun's (now Oracle) CDDL
license.  Original  copyrights  are left intact. No GPL code is
incorporated into the release, to avoid legal conflicts.

Any  mistakes  or omissions in copyright attribution will be my
mistake, so please let me know if there are such cases.

The  linux kernel was referred to in order to engineer the glue
for  dtrace  behavior, and there is no intention of making this
code fall under anything other than CDDL. (If Oracle migrate to
a GPL friendly license, then this port of dtrace can follow). I
do  not  own  the license or assert any rights on the licensing
other than that expected of me as a consumer/supplier.

I  have  no political affiliation or preference for a licensing
scheme,  simply  that  Sun/Oracle has gracefully donated to the
community a large body of work.

I  reserve  the  right to change the licensing model for my own
code  at  a later date, when and if someone puts forward a case
as to the correct license agreement.

If  the code is useful to you - great. Spread it around and get
people to use, debug and enhance it.

GIT Repository
==============

  https://github.com/dtrace4linux/linux

Installation
============

You may need to grab some extra packages for building DTrace.
Use either of the following to download extra packages. This
list may be incomplete depending on the version of your kernel/distro.

	$ tools/get-deps-arch.sh	# if using ArchLinux
	$ tools/get-deps.pl 		# if using Ubuntu
	$ tools/get-deps-fedora.sh	# RedHat/Fedora

	$ make all
	$ make install
	$ make load           (need to be root or have sudo access)

If the libdwarf package installed on the system is to old
it still compiles without any problem, but you will get
runtime errors from the io.d and/or sched.d files due to
undefined kernel structure definitions.

If you get a undefined struct definition such as dtrace_cpu_t
when running, please upgrade it.

Tested successfully with version 20100214 (whereas 20080409
is to old).

More details
============

Building is done in a build/ directory. The makefiles allow
you to compile for alternate kernel releases in the same tree,
which is useful for cross-version checking.

The result is:

	build/dtrace               User land executable
	build/drti.o               Object file for USDT apps
	build/driver/dtracedrv.ko  Kernel loadable module

Installing will copy them to Solaris compliant locations:

	/usr/sbin/dtrace
	/usr/lib/dtrace/64/drti.o

You dont need to 'install' to run dtrace, but you will need
to load the driver.

Kernel versioning
=================
dtrace relies on a kernel module and so a binary is needed
per system you deploy to, or kernel version.

dtrace is sensitive to the kernel - and attempts to cater for that,
but very old, or very new kernels may not have been validated.
Please feed back if that is the case.

A kernel strack trace is expected when loading the module, due to
currently unknown reasons (the current theory is that the kernel
ftrace mechanism which probes dtrace as its loaded gets confused
by what it sees).

If you get a stack trace something like this in the logs when
loading the module, this can safely be ignore (it appears to
be no harm (unless you use system-tap and dtrace at the same
time, then there could be a conflict):

    [  182.556392] dtracedrv: module license 'CDDL' taints kernel.
    [  182.556396] Disabling lock debugging due to kernel taint
    [  184.760136] CPU: 5 PID: 11008 Comm: dtrace Tainted: P           O 3.12.0+scst+tf.1 #5
    [  184.760140] Hardware name: To be filled by O.E.M. To be filled by O.E.M./SABERTOOTH 990FX, BIOS 0901 11/24/2011
    [  184.760142]  ffffffffa093c8a0 ffffffff813c18b9 ffff8800dacdaa80 ffffffffa08f28ed
    [  184.760146]  755f6f745f646162 ffff8800dacdaa80 ffff88040acbbb9c 0000000000000020
    [  184.760149]  0000000000000001 ffffffffa09370c0 ffff8800dd020a80 ffffffffa08e7f3e
    [  184.760151] Call Trace:
    [  184.760156]  [<ffffffff813c18b9>] ? dump_stack+0x41/0x58
    [  184.760165]  [<ffffffffa08f28ed>] ? mutex_enter_common+0x2d/0xeb [dtracedrv]
    [  184.760172]  [<ffffffffa08e7f3e>] ? par_alloc+0x20/0xd0 [dtracedrv]
    [  184.760178]  [<ffffffffa08f0894>] ? instr_provide_module+0x31/0x1f5 [dtracedrv]
    [  184.760184]  [<ffffffffa08f489c>] ? sdt_open+0x3/0x3 [dtracedrv]
    [  184.760189]  [<ffffffffa08f2980>] ? mutex_enter_common+0xc0/0xeb [dtracedrv]
    [  184.760195]  [<ffffffffa08d83d0>] ? dtrace_probe_provide+0xcd/0xf7 [dtracedrv]
    [  184.760201]  [<ffffffffa08d844f>] ? dtrace_open+0x55/0x10f [dtracedrv]
    [  184.760203]  [<ffffffff812b3e60>] ? kobj_lookup+0xfc/0x133
    [  184.760209]  [<ffffffffa08e7400>] ? dtracedrv_open+0x4c/0x51 [dtracedrv]
    [  184.760212]  [<ffffffff812a3973>] ? misc_open+0x107/0x168
    [  184.760216]  [<ffffffff8112c091>] ? chrdev_open+0x129/0x148
    [  184.760218]  [<ffffffff8112bf68>] ? cdev_put+0x1a/0x1a
    [  184.760220]  [<ffffffff81127082>] ? do_dentry_open+0x16c/0x22b
    [  184.760221]  [<ffffffff8112720f>] ? finish_open+0x2c/0x35
    [  184.760224]  [<ffffffff81134d9d>] ? do_last+0x9fd/0xc4a
    [  184.760226]  [<ffffffff81135249>] ? path_openat+0x25f/0x5bd
    [  184.760228]  [<ffffffff81141561>] ? mntput_no_expire+0x1b/0x16c
    [  184.760230]  [<ffffffff811356a2>] ? do_filp_open+0x2d/0x75
    [  184.760233]  [<ffffffff8111e32c>] ? kmem_cache_alloc+0x114/0x194
    [  184.760235]  [<ffffffff813c4d93>] ? _raw_spin_unlock+0x9/0xb
    [  184.760237]  [<ffffffff8113fa09>] ? __alloc_fd+0xfa/0x10c
    [  184.760239]  [<ffffffff81126cfa>] ? do_sys_open+0x146/0x1d6
    [  184.760241]  [<ffffffff813c7909>] ? ia32_do_call+0x13/0x13

No Linux Kernel source modifications required
=============================================

This is important for a number of reasons -- unless dtrace
is accepted into the kernel, it has to live with changes to header
files and data structures. Also, from a licensing perspective it
is not valid for dtrace to touch your sources. It is also much
easier to not even require kernel sources - so long as 
a kernel build environment is available.

INSTALLATION
============

Run 'make' with no arguments to see the current options. You
may need to run one of the tools/get-deps scripts for your OS
flavor to ensure you have the tools and kernel build
environment for your kernel.

make all
	to compile the drivers and user space commands. Check the file
	Packages, for hints on what you need (not much, but libelf, kernel
	source, flex/yacc -- bison will do).

make install
	Copy dtrace binary and driver to correct install location.

make load
	To load the drivers, and then you can play with cmd/dtrace/dtrace.

make unl
	to unload the drivers.

make test
	To run the userland cmd/dtrace regression test

To build the userland (command and object file etc) and the
kernel module for different architectures, set the environment
variable BUILD_ARCH appropriately and then use the make targets
separately.

This example is for building on a system with a 64-bit kernel,
but with 32-bit userland:

       BUILD_ARCH=i386 make cmds
       BUILD_ARCH=x86_64 make kernel

Dependencies
============
To build dtrace for linux requires a number of tools - mostly
the basic Unix development tools, plus you will need the kernel
source/build tree. dtrace does not affect or touch your kernel
sources, but it needs the normal header files for creating a 
loadable module.

Examine the following scripts to help identify missing packages:

	tools/get-deps-arch.sh
	tools/get-deps-fedora.sh
	tools/get-deps.pl

Internet scripts
================
Many scripts on the 'Net won't work since they tend to assume a
Solaris kernel, but if you look at them and read them to learn,
then they can mostly be adapted for Linux.


Changes
Mon Jan 11 21:17:16 2016  fox

     868* driver/intr.c
        Handle 3.10 (centos 7) kernels for store_gdt.

Mon Dec 28 19:47:46 2015  fox

     867* driver/dtrace.c
          driver/dtrace_linux.c
          driver/fasttrap_isa.c
          Fixes for fasttrap (pid provider):

	  dtrace -n pidXXXX:::entry

	  will work but *WILL* cause any fork/cloned child to terminate
	  with a SIGTRAP event because we dont latch onto the children
	  and remove the embedded breakpoints. (Needs more R&D to determine
	  how to do this in Linux).

	  This should work, but if it fails, send me a dmesg output -
	  the patching of the page tables may/will fail when allocating
	  a private buffer - but this may depend on VM, Xen or real HW.

     866* README.md
          driver/intr_x86-64.S
          driver/systrace.c
          driver/systrace_asm.S
          driver/x_call.c
          libdtrace/drti.c
          libproc/common/Psymtab.c
          tools/load.pl
          tools/mkport.pl
          tools/mkrelease.pl
          Fix for system_unbound_workq vs system_unbound_wq.
	  Fix silly in Psymtab.c leading to SIGSEGV.

Tue May 19 21:20:33 2015  fox

     865* driver/prov_proc.c
         Fix for 3.18/19 kernels (not sure I got the change right - 3.16
	 is not affected, but 3.19 - Ubuntu 15.04 - is).
        
Tue Mar  3 22:33:03 2015  fox

     864* README
          Status.txt
          archive.sh
          cmd/ctfconvert/barrier.c
          cmd/ctfconvert/makefile
          driver/Makefile
          driver/ctl.c
          driver/dtrace.c
          driver/dtrace_asm.c
          driver/dtrace_linux.c
          driver/dtrace_linux.h
          driver/fbt_linux.c
          driver/instr_linux.c
          driver/intr.c
          driver/intr_x86-64.S
          driver/prov_proc.c
          driver/systrace.c
          driver/systrace_asm.S
          driver/taskq.c
          driver/toxic.c
          driver/x_call.c
          driver-kmem/Makefile
          driver-kmem/dtrace_kmem.c
          driver-kmem/mk
          libdtrace/dt_names.c
          makefile
          s/README
          s/all
          s/clone
          s/execve
          s/fork
          s/iopl
          s/rt_sigreturn
          s/sigaltstack
          s/sigreturn
          s/vfork
          s/vm86
          s/z
          s/z1
          s/z11
          s/z2
          s/z3
          s/z4
          s/z5
          s/z6
          s/z7
          s/z8
          s/z9
          tests/syscalls.c
          tests/tests.d
          tools/build.pl
          tools/kcore.c
          tools/load.pl
          tools/mkdriver.pl
          tools/mkport.pl
          tools/mksyscall.pl
          tools/sudo
          tools/tests.pl
          Archive: 1.0155

Wed Aug 20 20:59:48 2014  fox

     863* driver/systrace_asm.S
          Fix dup symbols for kernels < 3.7.

     862* cmd/ctfconvert/barrier.c
          cmd/ctfconvert/barrier.h
          cmd/ctfconvert/ctfmerge.c
          cmd/ctfconvert/makefile
          Properly build ctfmerge. We were generating two ctfconvert's.

Tue Aug 19 22:36:14 2014  fox

     861* cmd/ctfconvert/makefile
          Build ctfmerge.

Fri Aug 15 23:25:42 2014  fox

     860* driver/ctl.c
          Fix for 3.16.1 kernel and confusion of uid_t vs kuid_t.

Thu Aug  7 21:23:14 2014  fox

     859* .release
          Add missing driver/systrace_asm.S to the release.

Wed Aug  6 22:22:40 2014  fox

     858* .release
          Status.txt
          driver/systrace.c
	  Fix silly typo left in from debugging 3.7 kernels.

Sat Aug  2 11:36:57 2014  fox

     857* driver/Makefile
          driver/systrace.c
          driver/systrace_asm.S
          tests/syscalls.c
          tools/load.pl
          Fix the ptregs issues on >= 3.7 kernels. sys_execve/x32 has
	  moved to compat_sys_execve.

Sun Jun  8 13:52:17 2014  fox

     856* .release
          README
          cmd/ctfconvert/makefile
          cmd/ctfconvert/strtab.c
          tools/get-deps.pl
	  Dont install openssh-server unless its "me".

Sat Apr 12 20:47:38 2014  fox

     855* 
          /home/fox/src/dtrace/driver/dtrace_asm.c
        dtrace_asm.c: Avoid arch_local_irq_save/restore due to older kernels.

Sun Mar 23 21:57:52 2014  fox

     854* .release
          archive.sh
          Archive: 1.0153

     853* .release
          README
          driver/dtrace_asm.c
          driver/dtrace_linux.c
          driver/mutex.c
          driver/prov_common.c
          driver/xen.c
          libdtrace/dt_proc_linux.c
          librtld/rtld_db.c
          linux/rtld_db.h
          tools/get-deps.pl
        dtrace_asm.c: Avoid CLI/STI to avoid issue on Xen kernels. 
	   (Ubuntu 12.04).
	Minor changes elsewhere - dt_proc_linux.c is an attempt to 
	make the "-c" and pid provider work when we launch a
	process but its not yet finished.

Fri Jan  3 16:11:59 2014  fox

     852* driver/dtrace_linux.c, driver/dtrace_linux.h, driver/prov_proc.c:
         Additional uid_t/kuid_t fixes.

     851* driver/ctl.c, driver/dtrace.c, driver/dtrace_linux.h:
          More fixes for uid_t/kuid_t.

     850* driver/ctl.c: Fix for 3.8.0 kernel kuid_t

Sun Nov  3 14:34:37 2013  fox

     849* Archive: 1.0152

     848* tools/readelf.pl: New/experimental script

     847* driver/Makefile: Use -gdwarf-2 debug info so we work on
          Ubuntu 13.10/gcc-4.8 (else mkctf.sh dies).

Sat Nov  2 18:14:37 2013  fox

     846* ctfconvert/dwarf.c: Silly hack for DWARF-4 issue. I still cant
          figure out libdwarf vs libdw and how they not only evolve but
	  the features dont allow one to be implemented from another.
	  This is a fricking mess.

Sun Oct 20 10:48:45 2013  fox

     845* Archive: 1.0151

Sat Oct 19 17:47:23 2013  fox

     844* driver/dtrace_linux.c: Fix for /proc/dtrace/trace on 3.11 kernel
          which doesnt like big blobs of output.

Fri Oct 18 22:51:33 2013  fox

     843* driver/systrace.c: Dont GPF on an amd64 kernel with no
          32-bit syscalls support when catting /proc/dtrace/syscall.


etc/regs.d
/*
GPLv3 license. This is a generated file from mkinstall.pl.

This file is part of the Dtrace/Linux distribution.
Generated: 20100728 12:38:34

$Header: Last edited: 28-Jul-2010 1.1 $ 

*/

inline int R_GS = 0;
#pragma D binding "1.0" R_GS
inline int R_FS = 1;
#pragma D binding "1.0" R_FS
inline int R_ES = 2;
#pragma D binding "1.0" R_ES
inline int R_DS = 3;
#pragma D binding "1.0" R_DS

inline int R_EDI = 4;
#pragma D binding "1.0" R_EDI
inline int R_ESI = 5;
#pragma D binding "1.0" R_ESI
inline int R_EBP = 6;
#pragma D binding "1.0" R_EBP
inline int R_ESP = 7;
#pragma D binding "1.0" R_ESP
inline int R_EBX = 8;
#pragma D binding "1.0" R_EBX
inline int R_EDX = 9;
#pragma D binding "1.0" R_EDX
inline int R_ECX = 10;
#pragma D binding "1.0" R_ECX
inline int R_EAX = 11;
#pragma D binding "1.0" R_EAX

inline int R_TRAPNO = 12;
#pragma D binding "1.0" R_TRAPNO
inline int R_ERR = 13;
#pragma D binding "1.0" R_ERR
inline int R_EIP = 14;
#pragma D binding "1.0" R_EIP
inline int R_CS = 15;
#pragma D binding "1.0" R_CS
inline int R_EFL = 16;
#pragma D binding "1.0" R_EFL
inline int R_UESP = 17;
#pragma D binding "1.0" R_UESP
inline int R_SS = 18;
#pragma D binding "1.0" R_SS

inline int R_PC = R_EIP;
#pragma D binding "1.0" R_PC
inline int R_SP = R_UESP;
#pragma D binding "1.0" R_SP
inline int R_PS = R_EFL;
#pragma D binding "1.0" R_PS
inline int R_R0 = R_EAX;
#pragma D binding "1.0" R_R0
inline int R_R1 = R_EBX;
#pragma D binding "1.0" R_R1

/* amd64/64-bit registers. */

inline int R_RDI = 0;
#pragma D binding "1.0" R_RDI
inline int R_RSI = 1;
#pragma D binding "1.0" R_RSI
inline int R_RDX = 2;
#pragma D binding "1.0" R_RDX
inline int R_RCX = 3;
#pragma D binding "1.0" R_RCX
inline int R_R8 = 4;
#pragma D binding "1.0" R_R8
inline int R_R9 = 5;
#pragma D binding "1.0" R_R9
inline int R_RAX = 6;
#pragma D binding "1.0" R_RAX
inline int R_RBX = 7;
#pragma D binding "1.0" R_RBX
inline int R_RBP = 8;
#pragma D binding "1.0" R_RBP
inline int R_R10 = 9;
#pragma D binding "1.0" R_R10
inline int R_R11 = 10;
#pragma D binding "1.0" R_R11
inline int R_R12 = 11;
#pragma D binding "1.0" R_R12
inline int R_R13 = 12;
#pragma D binding "1.0" R_R13
inline int R_R14 = 13;
#pragma D binding "1.0" R_R14
inline int R_R15 = 14;
